util/topgen/c.py

```diff
@@ -30,27 +49,70 @@ def __init__(self, top_name, name, base_addr, size_bytes):
         self.size_bytes = size_bytes
 
     def base_addr_name(self):
-        return self.name.with_more(prefix=["top", self.top_name],
+        return self.name.with_more(prefix=self.top_name.parts,
                                    suffix=["base", "addr"])
 
     def size_bytes_name(self):
-        return self.name.with_more(prefix=["top", self.top_name],
+        return self.name.with_more(prefix=self.top_name.parts,
                                    suffix=["size", "bytes"])
 
+class CEnum(object):
+    def __init__(self, top_name, enum_name):
+        self.enum_name = Name.join(top_name, enum_name)
+        self.enum_counter = 0
+        self.finalized = False
+
+        self.constants = []
+
+    def add_constant(self, constant_name, value=None, docstring=""):
+        assert not self.finalized
+
+        full_name = Name.join(self.enum_name, constant_name)
+        if value is None:
+            value = self.enum_counter
+            self.enum_counter += 1
+
+        self.constants.append( (full_name, value, docstring) )
+
+    def add_last_constant(self, docstring=""):
+        assert not self.finalized
+
+        full_name = self.enum_name.with_more(suffix=["last"])
+
+        _, last_val, _ = self.constants[-1]
+
+        self.constants.append( (full_name, last_val, "\internal " + docstring ) )
+        self.finalized = True
+
+    @property
+    def name(self):
+        return self.enum_name
```

Comment:
Is there a reason not to just assign the name to `self.name`? Then you wouldn't need the property at all (and this is a `CEnum`, so the "enum" part of `enum_name` seems a bit superfluous).